controllers/reconcilers/kserve_isvc_gateway_reconciler.go

```diff
@@ -0,0 +1,263 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reconcilers
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	kservev1beta1 "github.com/kserve/kserve/pkg/apis/serving/v1beta1"
+	"github.com/opendatahub-io/odh-model-controller/controllers/comparators"
+	"github.com/opendatahub-io/odh-model-controller/controllers/constants"
+	"github.com/opendatahub-io/odh-model-controller/controllers/processors"
+	"github.com/opendatahub-io/odh-model-controller/controllers/resources"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	istiov1beta1 "istio.io/api/networking/v1beta1"
+	istioclientv1beta1 "istio.io/client-go/pkg/apis/networking/v1beta1"
+)
+
+const (
+	internalSerivceHostnameDomain = "svc.cluster.local"
+	opendatahubManagedLabelName   = "opendatahub.io/managed='true'"
+	secretCheckInterval           = 1 * time.Second
+	maxSecretCheckAttempts        = 5
+)
+
+var _ SubResourceReconciler = (*KserveGatewayReconciler)(nil)
+
+type KserveGatewayReconciler struct {
+	client         client.Client
+	secretHandler  resources.SecretHandler
+	gatewayHandler resources.GatewayHandler
+	deltaProcessor processors.DeltaProcessor
+}
+
+func NewKserveGatewayReconciler(client client.Client) *KserveGatewayReconciler {
+	return &KserveGatewayReconciler{
+		client:         client,
+		secretHandler:  resources.NewSecretHandler(client),
+		gatewayHandler: resources.NewGatewayHandler(client),
+		deltaProcessor: processors.NewDeltaProcessor(),
+	}
+}
+
+func (r *KserveGatewayReconciler) Reconcile(ctx context.Context, log logr.Logger, isvc *kservev1beta1.InferenceService) error {
+	log.V(1).Info("Reconciling KServe local gateway for Kserve InferenceService")
+
+	// return if URL is not set
+	if isvc.Status.URL == nil {
+		log.V(1).Info("Waiting for the URL as the Inference Service is not ready yet")
+		return nil
+	}
+
+	destinationSecretExist := true
+	if _, err := r.secretHandler.Get(ctx, types.NamespacedName{Name: isvc.Name, Namespace: constants.IstioNamespace}); err != nil {
+		if errors.IsNotFound(err) {
+			destinationSecretExist = false
+		} else {
+			return err
+		}
+	}
+
+	if !destinationSecretExist {
+		secret := &corev1.Secret{}
+		for attempt := 1; attempt <= maxSecretCheckAttempts; attempt++ {
```

Comment:
isn't it safer to retriger the reconcile loop?

Reply:
That's a good question. There are some time gaps between creating a Secret for isvc and calling this gateway Reconciler. So I think it would be better to retry 5 times (but I expect only 1~2 times will happen) to wait for the secret creation.  @israel-hdez what do you think?

Reply:
when calling the reconciler loop, IIRC we can set the wait time to run again, we could have a counter, if that reaches, we give up.

Reply:
updated